Zstd support

mehmet-karaman opened the pull request:
Implemented Zstd support upon the Apache commons compress zip archive implementation.

mehmet-karaman committed:
Added Zstd support

mehmet-karaman committed:
Added zstd-jni library support

iloveeclipse committed:
Cleanup of bundle dependencies and versions

- fixed wrong "wrapped" m2e bundle dependency on airlift in tests
- fixed wrong classpath entry for aircompressor in zipeditor
- fixed project settings to refer to proper Java 21 environment
- bumped zipeditor bundle/feature/update site version to 1.2.0

mehmet-karaman committed:
- Extracted the ZSTD Library usages,
- moved to a central utility
- added handlers to handle these library usages. This helps to avoid the
ClassNotFoundExceptions
- Added message in Zstd Preferences Page to show if the Library is
available or not.

iloveeclipse commented on ZipEditor/src/zipeditor/model/ZipModel.java:
```diff
@@ -151,6 +170,10 @@ public static ContentTypeId detectType(InputStream contents) {
 
 		} catch (IOException e) {
+			if (e instanceof ZipEditorZstdException) {
+				ErrorDialog.openError(Display.getDefault().getActiveShell(), "Zstd in Zip File Detected", e.getMessage(), ZipEditorPlugin.log(e));;
```
`Zstd in Zip File Detected` is not a helpful end user message. Is it good? Is it bad? Why is this a problem? The message should help to understand what is the problem and give a hint why it happened. Here the problem would be something like `Error reading zip archive with Zstd encoding`.

Also, since this is code that might be executed multiple times, it is probably better to use status handler instead (it will group multiple errors in one dialog, not opening multiple).

`StatusManager.getManager().handle(status, StatusManager.SHOW | StatusManager.LOG);`
That will log **and** show an error dialog.

Beside this, there is a duplicated semicolon at the end of the line.

mehmet-karaman replied:
done, thanks for this hint..

iloveeclipse commented on ZipEditor/src/zipeditor/model/ZipModel.java:
```diff
@@ -295,18 +318,20 @@ private void readStream(InputStream zipStream, IModelInitParticipant participant
-					message += " " + zipPath.getAbsolutePath(); //$NON-NLS-1$
-				logError(ZipEditorPlugin.createErrorStatus(message, e));
-				break;
+				if (!(e instanceof ZipEditorZstdException)) {
```
So we will continue with the code below without reporting an error? Why?
This should be documented in code, it is not obvious.

Also I'm not sure if you really want the loop to continue?

mehmet-karaman replied:
Yes .. it didn't make sense to exclude zcheck for ZipEditorZstdException anymore.. reverted that code..

iloveeclipse commented on ZipEditor/src/zipeditor/model/ZipModel.java:
```diff
@@ -525,17 +555,27 @@ private void saveNode(OutputStream out, Node node, ContentTypeId type, IProgress
 			((TarOutputStream) out).putNextEntry(tarEntry);
-		Utils.readAndWrite(node.getContent(), out, false);
+		if (node instanceof ZipNode && ((ZipNode)node).getMethod() == ZipMethod.ZSTD.getCode()) {
+			OutputStream zstdOutput = ZstdUtilities.getOutputStream(out);
```
`close()` is really needed here (before it wasn't)? If yes, should be in the finally block. Otherwise there seem to be `close()` in the caller method that does this

mehmet-karaman replied:
changed that to try with resources. I guess the flush was necessary part.. (both libraries tried and both work with that)

iloveeclipse commented on ZipEditor/src/zipeditor/operations/AddOperation.java:
```diff
@@ -82,14 +84,14 @@ public IStatus runInUIThread(IProgressMonitor monitor) {
 
 	public void execute(String[] fileNames, Node parentNode, Node beforeSibling, StructuredViewer viewer) {
-		execute(getFilesFromNames(fileNames), parentNode, beforeSibling, viewer);
+		execute(getFilesFromNames(fileNames), parentNode, beforeSibling, viewer, false);
```
`false` is hard coded here? Why? Please add comment in the code if this is expected.

mehmet-karaman replied:
Default case is to not allow zstd compression. The zstd compression is only allowed for zip files. added a comment.

iloveeclipse reviewed (changes requested):
1) If the zstd support doesn't work or disabled, editor opens and is empty etc.
So far so good. But editor should not allow to save or change anything, right now it allows to add entries, it will be dirty after that but fails to save via NPE:

```
java.lang.NullPointerException: Cannot invoke "org.eclipse.core.runtime.content.IContentType.getFileSpecs(int)" because the return value of "org.eclipse.core.runtime.content.IContentTypeManager.getContentType(String)" is null
	at zipeditor.model.ZipContentDescriber.getContentTypeForFileExtension(ZipContentDescriber.java:99)
	at zipeditor.ZipEditor.internalSave(ZipEditor.java:384)
	at zipeditor.ZipEditor.doSave(ZipEditor.java:364)
```
Ideally editor would disable all actions, or, if it is toomuch work, at least do not fail with NPE but with a message that the file can't be changed.


2) If the zip can't be opened (zstd is disabled) and also sometimes if it is enabled (so it seem to be existing issue) I get following errors in the log:
```
java.lang.NullPointerException: Cannot invoke "zipeditor.ZipOutlinePage.setInput(zipeditor.model.Node)" because "this.this$0.fOutlinePage" is null
	at zipeditor.ZipEditor$19.run(ZipEditor.java:1125)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5047)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4527)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1151)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1042)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:668)
```

Fix:
```
diff --git a/ZipEditor/src/zipeditor/ZipEditor.java b/ZipEditor/src/zipeditor/ZipEditor.java
index 5561f2c..1869e05 100644
--- a/ZipEditor/src/zipeditor/ZipEditor.java
+++ b/ZipEditor/src/zipeditor/ZipEditor.java
@@ -549,2 +549,3 @@
 		fZipActionGroup.dispose();
+		fModel = null;
 		super.dispose();		
@@ -1124,3 +1125,5 @@
 				public void run() {
-					fOutlinePage.setInput(fModel.getRoot());
+					if(fOutlinePage != null && fModel != null) {
+						fOutlinePage.setInput(fModel.getRoot());
+					}
 				}
```

3) Right clicking on a zip file with zstd disabled shows an error popup with the message that zstd support is not there. I believe this error popup is not needed here, logging should be enough.
Stack:
 ```
zipeditor.model.ZipEditorZstdException: Support for zstd is deactivated. Please check the zip editor settings.
	at zipeditor.model.zstd.ZstdUtilities.getInputStream(ZstdUtilities.java:16)
	at zipeditor.model.ZipModel$1.createZstdInputStream(ZipModel.java:115)
	at org.apache.commons.compress.archivers.zip.ZipArchiveInputStream.getNextZipEntry(ZipArchiveInputStream.java:791)
	at org.apache.commons.compress.archivers.zip.ZipArchiveInputStream.getNextEntry(ZipArchiveInputStream.java:648)
	at zipeditor.model.ZipModel.detectType(ZipModel.java:119)
	at zipeditor.model.ZipContentDescriber.describe(ZipContentDescriber.java:144)
	at org.eclipse.core.internal.content.ContentTypeCatalog.describe(ContentTypeCatalog.java:295)
	at org.eclipse.core.internal.content.ContentTypeCatalog.collectMatchingByContents(ContentTypeCatalog.java:266)
	at org.eclipse.core.internal.content.ContentTypeCatalog.internalFindContentTypesForSubset(ContentTypeCatalog.java:490)
	at org.eclipse.core.internal.content.ContentTypeCatalog.internalFindContentTypesFor(ContentTypeCatalog.java:549)
	at org.eclipse.core.internal.content.ContentTypeCatalog.findContentTypesFor(ContentTypeCatalog.java:383)
	at org.eclipse.core.internal.content.ContentTypeMatcher.findContentTypesFor(ContentTypeMatcher.java:57)
	at org.eclipse.lsp4e.LSPEclipseUtils.getFileContentTypes(LSPEclipseUtils.java:1327)
	at org.eclipse.lsp4e.LanguageServersRegistry.canUseLanguageServer(LanguageServersRegistry.java:463)
	at org.eclipse.lsp4e.HasLanguageServerPropertyTester.test(HasLanguageServerPropertyTester.java:29)
	at org.eclipse.core.internal.expressions.Property.test(Property.java:65)
	at org.eclipse.core.expressions.TestExpression.evaluate(TestExpression.java:107)
	at org.eclipse.core.expressions.CompositeExpression.evaluateAnd(CompositeExpression.java:54)
	at org.eclipse.core.internal.expressions.AdaptExpression.evaluate(AdaptExpression.java:121)
	at org.eclipse.core.expressions.CompositeExpression.evaluateOr(CompositeExpression.java:68)
	at org.eclipse.core.expressions.OrExpression.evaluate(OrExpression.java:26)
	at org.eclipse.core.expressions.CompositeExpression.evaluateAnd(CompositeExpression.java:54)
	at org.eclipse.core.internal.expressions.IterateExpression.evaluate(IterateExpression.java:163)
	at org.eclipse.e4.ui.internal.workbench.ContributionsAnalyzer.isCoreExpressionVisible(ContributionsAnalyzer.java:263)
	at org.eclipse.e4.ui.internal.workbench.ContributionsAnalyzer.isVisible(ContributionsAnalyzer.java:237)
	at org.eclipse.e4.ui.workbench.renderers.swt.ContributionRecord.computeVisibility(ContributionRecord.java:167)
	at org.eclipse.e4.ui.workbench.renderers.swt.ContributionRecord.updateVisibility(ContributionRecord.java:98)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRendererFilter.updateElementVisibility(MenuManagerRendererFilter.java:173)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRendererFilter.updateElementVisibility(MenuManagerRendererFilter.java:184)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerShowProcessor.showMenu(MenuManagerShowProcessor.java:248)
```

4) If zstd support is disabled, I expect "usual" zip files to work. However, creating a new empty zip file, adding few entries and save results in an error. Same happens if I open existing regular zip file, add an entry and try to save:
```
java.lang.NullPointerException: Cannot invoke "org.eclipse.core.runtime.content.IContentType.getFileSpecs(int)" because the return value of "org.eclipse.core.runtime.content.IContentTypeManager.getContentType(String)" is null
	at zipeditor.model.ZipContentDescriber.getContentTypeForFileExtension(ZipContentDescriber.java:99)
	at zipeditor.ZipEditor.internalSave(ZipEditor.java:384)
	at zipeditor.ZipEditor.doSave(ZipEditor.java:364)
```
That should not happen.

iloveeclipse commented:
Another observation: starting without jnilib and aircompressor: the preference page doesn't disable the comboand the checkbox for zstd support is enabled and checked. Should be all disabled and some error label shown like "to enable zstd install jnilib or aircompressor libraries"

mehmet-karaman commented:
> Another observation: starting without jnilib and aircompressor: the preference page doesn't disable the comboand the checkbox for zstd support is enabled and checked. Should be all disabled and some error label shown like "to enable zstd install jnilib or aircompressor libraries"

Good observation, as the libraries are optional, we have to set the default "zstd active" flag to false. And this should be cared by the user if he wants to activate this, he has to take care of the installed libraries..

mehmet-karaman committed:
- improved the handling of selected / available library usages
- active flag is used also with checking if any library is available
- fixed NPEs
- Added hint in zip editor if zstd isn't available.
- fix Project Explorer expanding bug (if afterwards the zstd was
deactivated)
- changed resource close() to try with resources.

mehmet-karaman commented:
Could reproduce the bug in the Package Explorer.. 

Zstd support was active, the expander besides the zip elements tree item was visible.. Disabled the Zstd Support and expandede that node and it has thrown the NPE.. Now this is also fixed..

iloveeclipse commented on ZipEditor/src/zipeditor/model/ZipModel.java:
```diff
@@ -102,10 +106,24 @@ public static ContentTypeId detectType(InputStream contents) {
+				if (ioe instanceof ZipEditorZstdException) {
+					return ContentTypeId.INVALID;
+				}
+				// thrown in getNextEntry() method if, if file is not a zip file.
```
Not sure I understood this change & comment.
So if the exception is **not** `ZipEditorZstdException` we **silently continue**? Shouldn't we re-throw `IOException` here, or, better, catch `ZipEditorZstdException` only?

mehmet-karaman replied:
Yes.. It tries to identify the next archive type.. But if the ZipEditorZstdException is thrown, we know that we can't open the archive anymore (due to missing library or deactivated zstd support). See line 140 where the next archive check was missed and it is about to try the next archive type.

Added a comment and fixed the comment in line 126.

iloveeclipse commented on ZipEditor/src/zipeditor/model/ZipModel.java:
```diff
@@ -151,6 +169,11 @@ public static ContentTypeId detectType(InputStream contents) {
 			return ContentTypeId.TAR_FILE;
 
 		} catch (IOException e) {
+			if (e instanceof ZipEditorZstdException) {
```
Is this a dead code? Isn't the exception ignored above?

mehmet-karaman replied:
The exception is catched above and returns already an invalid content type. It is dead code, which i am going to remove now.

iloveeclipse commented on ZipEditor/src/zipeditor/model/ZipModel.java:
```diff
@@ -428,7 +454,12 @@ private InputStream detectStream(InputStream contents) throws IOException {
 			return in;
 		case ContentTypeId.ZIP:
-			return new ZipInputStream(in);
+			return new ZipArchiveInputStream(in) {
```
I see this anonymous class is created in both places where `ZipArchiveInputStream` instance is created. I would create `ZsdtAwareZipArchiveInputStream` that overrides the methoid and create it instead in both places.

mehmet-karaman replied:
done.

iloveeclipse commented on ZipEditor/src/zipeditor/model/ZipNode.java:
```diff
@@ -83,10 +87,11 @@ protected InputStream doGetContent() throws IOException {
-			if (root.getZipFile() != null)
-				return root.getZipFile().getInputStream(zipEntry);
-			return model.getZipPath() != null ? new EntryStream(zipEntry, new ZipFile(model.getZipPath())) : null;
+			if (zipEntry.getMethod() == ZipMethod.ZSTD.getCode() || zipEntry.getMethod() == ZipMethod.ZSTD_DEPRECATED.getCode()) {
```
Hard to read... Can we rewrite this block to 
```
	protected InputStream doGetContent() throws IOException {
		InputStream in = super.doGetContent();
		if (in != null) {
			return in;
		}
		if (zipEntry != null && model.getZipPath() != null) {
			ZipFile zipFile;
			if (isZstdEncoded(zipEntry)) {
				zipFile = ZstdUtilities.getZipFileBuilder().setFile(model.getZipPath()).get();
			} else {
				zipFile = ZipFile.builder().setFile(model.getZipPath()).get();
			}
			return new EntryStream(zipEntry, zipFile);
		}
		return null;
	}

	private static boolean isZstdEncoded(ZipArchiveEntry zipEntry) {
		return zipEntry.getMethod() == ZipMethod.ZSTD.getCode() || zipEntry.getMethod() == ZipMethod.ZSTD_DEPRECATED.getCode();
	}
```

mehmet-karaman replied:
looks better.. rewritten the code according to this.,

iloveeclipse commented on ZipEditor/src/zipeditor/model/ZipRootNode.java:
```diff
@@ -4,34 +4,12 @@
-	}
-
-	public Object accept(NodeVisitor visitor, Object argument) throws IOException {
-		// this is relevant when saving to a previously empty file
```
Just curious, why is this code not needed anymore?

mehmet-karaman replied:
It was nothing special doing anymore.. It was overriden (without the overridden annotation).. Might be enough explained?

iloveeclipse replied:
OK

iloveeclipse commented on ZipEditor/src/zipeditor/model/zstd/ZstdImplementationHandler.java:
```diff
@@ -0,0 +1,36 @@
+	default OutputStream noClose(OutputStream out) {
+		return new OutputStream() {
+			@Override
+			public void write(byte[] b) throws IOException {
```
this method does nothing differently to the super, why it is overridden?

mehmet-karaman replied:
It is just delegating to the inner outputstream :) .. but the close method is not delegating..

iloveeclipse commented on ZipEditor/src/zipeditor/model/zstd/ZstdImplementationHandler.java:
```diff
@@ -0,0 +1,36 @@
+			}
+			
+			@Override
+			public void write(byte[] b, int off, int len) throws IOException {
```
this method does nothing differently to the super, why it is overridden?

mehmet-karaman replied:
The OutputStream Constructor doesn't have an argument so the constructor won't call its super outputstream with that inner outputstream. If the stream is closed, the inner stream wont be closed. 

The inner stream is still necessary for further stuff at least reading the zip central directory entry or a next entry if it has..

iloveeclipse commented on ZipEditor/src/zipeditor/preferences/PreferenceUtils.java:
```diff
@@ -0,0 +1,106 @@
+	 */
+	public static final String JNI_LIBRARY = "jniLibrary"; //$NON-NLS-1$
+
+	public static final String[][] libs = { { "JNI Library", JNI_LIBRARY }, { "Aircompressor", AIRCOMPRESSOR } }; //$NON-NLS-1$//$NON-NLS-2$
```
Can we please get rid of double dimensional array and use a list of records `ZstdLibrary(String label, String preferenceKey)`? This would be self explaining code compared to the current code below
```
if (selectedZstdLib.equals(strings[1])) {
 return strings[1];
}
```

mehmet-karaman replied:
This is used in the Preferences Combobox EditField .. I wasn't aware about that it also can handle that..

iloveeclipse commented on ZipEditor/src/zipeditor/preferences/PreferenceUtils.java:
```diff
@@ -0,0 +1,106 @@
+	 * Returns the available libraries. its a two dimensional array as the libs above but contains only the available ones.
+	 * @return the array of available libraries.
+	 */
+	public static String[][] getAvailableLibraries() {
```
The method is unreadable, it should be rewritten using records or enums.

mehmet-karaman replied:
As far as i remember we discussed this.. It was because it needs to deliver a two dimensional array with the Combobox entry label and value.

iloveeclipse commented on ZipEditor/src/zipeditor/preferences/PreferenceUtils.java:
```diff
@@ -0,0 +1,106 @@
+	 * 
+	 * @return the selected or available library.
+	 */
+	public static String getSelectedOrAvailableLibrary() {
```
It is absolutely unclear what is returned - a library path? Symbolic name? Label for UI? No idea.

mehmet-karaman replied:
Added additional explanation for the return value.. Its just an identifier which is coded in the constants JNI_LIBRARY or AIRCOMPRESSOR 

The additional comment
identifier of the library which is written in the constants JNI_LIBRARY or AIRCOMPRESSOR or null if there is no library available.

iloveeclipse commented on ZipEditor/src/zipeditor/preferences/ZipEditorPreferencePage.java:
```diff
@@ -0,0 +1,112 @@
+	
+	@Override
+	public void dispose() {
+		super.dispose();
```
Usual order of dispose it the order **opposite** to the construction where the super code is executed first. So here super.dispose() should be the last call.  If that is impossible, a comment should be added why the order is unusual one.

mehmet-karaman replied:
It just deregisters the listener from the preference store, which doesn't depened on the UI.. But i changed the order..

iloveeclipse commented on ZipEditor/src/zipeditor/preferences/messages.properties:
```diff
@@ -0,0 +1,6 @@
+ZipEditorPreferencePage_ActivateZstdSupport=Activate Zstd Support
+ZipEditorPreferencePage_HintNoLibraries=There are no Zstd libraries available.\nIf you want to use zstd compression, you have to install the libraries.
```
I would add here a hint **which** libraries are supported.
Something like 
```
Zstd zip compression support is not available.\nIf you want to use zstd compression, you have to install a library that provides zstd compression, like zstd-jni or aircompressor.
```

mehmet-karaman replied:
done.

mehmet-karaman committed:
- Added "is zstd default compression" preference
- Add new files by Drag And Drop with Zstd compression if the new pref
is true.
- Fixed Zip entry properties bug with zsdt in method combobox.
- Implemented PR comments.
- changed two dim. array of libraries (combobox edit field) to more
readable records and conversion to two dim. array.
- added checkbox selection of zstd default in dialog according the
default zstd compression state.
- added additional checks if zstd is used as default but this method is
not used in this zip file.

iloveeclipse commented on ZipEditor/src/zipeditor/preferences/messages.properties:
```diff
@@ -1,6 +1,9 @@
+ZipEditorPreferencePage_HintNoLibraries=Zstd zip compression support is not available.\nIf you want to use zstd compression, you have to install a library that provides zstd compression, like zstd-jni or aircompressor.
 ZipEditorPreferencePage_LibAvailable=Library is available
 ZipEditorPreferencePage_LibNotAvailable=Library is not available
+ZipEditorPreferencePage_ZstdAsDefaultCompression=ZSTD Default Compression
```
`ZSTD Default Compression` in the preferences below doesn't look good and sounds like an incomplete sentence:
![image](https://github.com/user-attachments/assets/3c355a51-2052-47af-b91d-b829b489fbaa)

It should be something like `Use Zstd compression by default for new zip files`.

iloveeclipse reviewed (changes requested):
Nnow I see that updating entries in existing *not-zstd* compressed file create zstd entries:
![image](https://github.com/user-attachments/assets/f06b9c1f-1b9e-42d6-99e1-f920f82cbb45)

The file below was created with default compression and .classpath inside was updated while the preference "ZSTD Default Compression" was enabled.

However, the preference should affect only new files, not existing.

Additionally to that, if such file is saved, next attempt to **open** it **with zstd disabled** results in an error message is shown "An error occurred trying to **save**":

![image](https://github.com/user-attachments/assets/d634c421-ac06-4b3a-95c2-86f65c017870)

After that the zip file **is opened** but the entry encoded with zstd compression is not shown (OK).

However, after that one could modify any other entry in the zip & save the zip, **loosing** the zstd encoded entry (NOK).

mehmet-karaman committed:
- Added different exception handling back.

mehmet-karaman committed:
Added Preference for Zstd compression level.

mehmet-karaman committed:
mavenify zip editor

mehmet-karaman committed:
mavenify zip editor (fix camel case problem)

mehmet-karaman committed:
fix wrong version number of zstd-jni bundle

mehmet-karaman committed:
negated initial logic for showing the scale and scale value controls.
Was checking before that JNI lib is selected, now its checking that
aircompressor is not selected.

uvoigt commented:
Hi,
I did not get notified by email about this activities. That's why I was so silent ;)
At the moment, I don't have the time to check the feature. But I do in a couple of weeks.
@iloveeclipse do you think the feature is complete?

iloveeclipse commented:
@uvoigt : at some point we assumed there is no one interested and most resent patches are not pushed here. We have few commits on top which clean up preferences page and "mavenize" the build to produce a fully self contained p2 repository with all dependencies included.
We wil push on Monday. With that it would be complete.

uvoigt commented:
Thanks @iloveeclipse,
I will have a look as soon as I get the time. I see that I pushed my latest patches up to the current release and I am not aware of other recent patches.
Anyway, thank you for your contribution and interest!
Uwe

mehmet-karaman committed:
fix line break issues in feature.xml's texts

mehmet-karaman commented:
Pushed the latest changes.
- Mavenified zip editor (uses tycho and eclipse orbit similar building) 
- fixed smaller ui glitches in the preference page


The maven build can be started with "mvn clean install". The built artifacts can be found in the following path "/zipeditor/ZipEditor.UpdateSite/target/repository"

uvoigt commented:
> Pushed the latest changes.
> 
> * Mavenified zip editor (uses tycho and eclipse orbit similar building)
> * fixed smaller ui glitches in the preference page
> 
> The maven build can be started with "mvn clean install". The built artifacts can be found in the following path "/zipeditor/ZipEditor.UpdateSite/target/repository"
Please give me some time @mehmet-karaman to have a look and test. I will definitely accept your changes at last. The only problem is that my time is somewhat limited these days.
Best regards,
Uwe

mehmet-karaman commented:
Hello @uvoigt ,

I've done the rebase, there are no conflicts anymore.

mehmet-karaman commented:
The maven build can be started with "mvn clean install". The built artifacts can be found in the following path "/zipeditor/ZipEditor.UpdateSite/target/repository"

After that the target platform can be selected as /ZipEditor/latest.target ..

mehmet-karaman committed:
- Fix problem with empty files and no local headers

uvoigt commented:
@mehmet-karaman 

when building I get

```
[INFO] [1m----< [0;36morg.eclipse.orbit.maven-bnd:org.eclipse.orbit.maven-bnd.site[0;1m >----[m
[INFO] [1mBuilding org.eclipse.orbit.maven-bnd.site 1.0.0-SNAPSHOT           [4/8][m
[INFO]   from maven-bnd/site/pom.xml
[INFO] [1m-------------------------[ eclipse-repository ]-------------------------[m
[INFO] Resolving dependencies of MavenProject: org.eclipse.orbit.maven-bnd:org.eclipse.orbit.maven-bnd.site:1.0.0-SNAPSHOT @ /Users/voigt/eclipse-workspace-memet/zipeditor/maven-bnd/site/pom.xml
[INFO] Resolving target definition file:/Users/voigt/eclipse-workspace-memet/zipeditor/maven-bnd/tp/MavenBND.target for environments=[win32/win32/x86_64, macosx/cocoa/x86_64, macosx/cocoa/aarch64, linux/gtk/x86_64, linux/gtk/aarch64], include source mode=honor, referenced repository mode =ignore, execution environment=StandardEEResolutionHints [executionEnvironment=OSGi profile 'JavaSE-21' { source level: 21, target level: 21}] with Tycho Provisioning Agent (extension>org.eclipse.tycho:tycho-maven-plugin:4.0.13)
[INFO] Resolving MavenDependencyRoots = [GroupId = io.airlift, ArtifactId = aircompressor, Version = 2.0.2, ArtifactType = jar], IncludeDependencyScope = [compile], MissingManifestStrategy = GENERATE, IncludeSource = true
[INFO] default instructions = {Export-Package=*;version="${version}";-noimport:=true, Bundle-SymbolicName=${mvnGroupId}.${mvnArtifactId}, Bundle-Version=${version}, Bundle-Name=Bundle derived from maven artifact ${mvnGroupId}:${mvnArtifactId}:${mvnVersion}, DynamicImport-Package=*, version=${version_cleanup;${mvnVersion}}, Import-Package=*;resolution:=optional}
[INFO] io.airlift:aircompressor:2.0.2 is wrapped as a bundle with bundle symbolic name io.airlift.aircompressor
[INFO] The artifact can be referenced in feature files with the following data: <plugin id="io.airlift.aircompressor" version="2.0.2"/>
[INFO] Resolving MavenDependencyRoots = [GroupId = org.apache.commons, ArtifactId = commons-compress, Version = 1.28.0-SNAPSHOT, ArtifactType = jar, GroupId = org.apache.commons, ArtifactId = commons-lang3, Version = 3.18.0-SNAPSHOT, ArtifactType = jar], IncludeDependencyScope = [compile], MissingManifestStrategy = GENERATE, IncludeSource = true
[INFO] Downloading from apache snapshot: https://repository.apache.org/content/repositories/snapshots/org/apache/commons/commons-compress/1.28.0-SNAPSHOT/maven-metadata.xml
[INFO] Downloading from apache snapshot: https://repository.apache.org/content/repositories/snapshots/org/apache/commons/commons-compress/1.28.0-SNAPSHOT/commons-compress-1.28.0-SNAPSHOT.pom
[WARNING] The POM for org.apache.commons:commons-compress:jar:1.28.0-SNAPSHOT is missing, no dependency information available
[INFO] Downloading from apache snapshot: https://repository.apache.org/content/repositories/snapshots/org/apache/commons/commons-compress/1.28.0-SNAPSHOT/commons-compress-1.28.0-SNAPSHOT.jar
[INFO] [1m------------------------------------------------------------------------[m
[INFO] [1mReactor Summary:[m
[INFO] 
[INFO] ZipEditor.parent 1.0.0-SNAPSHOT .................... [1;32mSUCCESS[m [  0.054 s]
[INFO] org.eclipse.orbit.maven-bnd.parent 1.0.0-SNAPSHOT .. [1;32mSUCCESS[m [  0.020 s]
[INFO] org.eclipse.orbit.maven-bnd.tp 1.0.0-SNAPSHOT ...... [1;32mSUCCESS[m [  2.064 s]
[INFO] org.eclipse.orbit.maven-bnd.site 1.0.0-SNAPSHOT .... [1;31mFAILURE[m [  0.335 s]
[INFO] ZipEditor 1.2.1-SNAPSHOT ........................... [1;33mSKIPPED[m
[INFO] Zip Editor Feature 1.2.1-SNAPSHOT .................. [1;33mSKIPPED[m
[INFO] ZipEditor.targetPlatform 1.0.0-SNAPSHOT ............ [1;33mSKIPPED[m
[INFO] ZipEditor.site 1.0.0-SNAPSHOT ...................... [1;33mSKIPPED[m
[INFO] [1m------------------------------------------------------------------------[m
[INFO] [1;31mBUILD FAILURE[m

```

Is it correct to depend on snapshots?

uvoigt commented:
I would like to have the maven build for the next release that should include the changes of `open-file-performance` and `master` + fallback implementation to `ZipArchiveInputStream.`
To achieve this, I simply copy the Maven files from here.
Or you prepare another PR with the mavenize commit. That might be easier.

Btw., I asked AI and it generates nearly the same stuff.

mehmet-karaman commented:
I would like to prepare a PR that contains the mavenization + ZipArchiveInputStream fallback..

uvoigt commented:
Today I had so much time. I added the maven things + Github action to the open-file-performance branch
It is already working with a new RC-release in the update manager.
Pardon if the code now differs from your solution. It seems to me, there are some files that are unnecessary.
Actually, the dependencies for commons-compress are not yet added. I will do when merging into master.

mehmet-karaman commented:
Ok, would you like to continue with the commons-compress integration (as fallback) or shall i continue from here? Do you have time to continue?

uvoigt commented:
Maybe tomorrow I have some time too. So I will try that.

mehmet-karaman commented:
Let me know if you need some help. Thanks.

uvoigt commented:
@mehmet-karaman I would offer to merge master into your branch because I caused the conflicts.
Not sure which would be the best way, e.g. I create a branch on the base commit of your branch and you retarget the PR. Then this can be fast-forwarded and I merge master into it. On top of that we can discuss.

I created fewer files when adding Maven support. It already runs in the Github CI so I want to stay with it. New release 1.2.0 is already out.
I would wish to integrate the Zstd support in smaller steps because there is potential for problems. I have seen that in the other PR.
Still, I am not familiar with the content of this PR. I gave up when I saw that the build failed.
Within the next days I probably will have time to work on it.

mehmet-karaman commented:
The build is currently failing due to unresolved dependencies on snapshot plugins, which have since been replaced by their release versions. To restore a working build, the dependencies need to be updated accordingly.

If you agree, I could rebase this branch on top of your master branch and resolve any resulting conflicts.

Additionally, I would like to ask whether, after including Zstd compression support, it would be possible to release a version higher than 1.2.1. This would allow us to drop our internal release and rely directly on your official release instead.

uvoigt commented:
> If you agree, I could rebase this branch on top of your master branch and resolve any resulting conflicts.

If you like, then try it. I have the following comments:

- What is the directory `maven-bnd` for? The master builds without it. If there is no good reason, I suggest to delete it.
- The master defines the target platform in the file `target-platform.target`. Since this works I don't see a need for the directory `ZipEditor.targetPlatform`. I suggest to delete it and to add the zstd dependencies there.


> Additionally, I would like to ask whether, after including Zstd compression support, it would be possible to release a version higher than 1.2.1. This would allow us to drop our internal release and rely directly on your official release instead.

I don't understand what you mean by 'our internal release'. From my point of view, the current release is 1.2.0. Therefore the next release is 1.2.1.
Btw. there is no need to define the release within the Maven descriptors. I use the Tycho Versions plugin to set the version from the Git tag.

uvoigt commented:
Answering https://github.com/uvoigt/zipeditor/pull/3#issuecomment-4074882518

>I've seen that minimal required is java 1.8, is there any problem to update this to java-17?

java 1.8, at least in JDT core preferences was chose by you within commit 7f2be9ed2c97a127803f6f1db880be7419392787
This actually is not fully consistent. I chose 2022-06 first within the target platform descriptor and modified it to 2022-09 later - but this not required at the moment
In general, from the beginning of this project on I focused on backwards compatibility because there were a lot of products based on Eclipse out there, especially within IBM. These products often were way behind actual Eclipse releases.
Therefore, generally I see no reason to break with that rule. Except for a good reason ;)

Please always state why you are planning something so that I can get an idea of ​​whether it is necessary or not.
E.g. which dependency requires Java 17?

At the moment, the picture is as follows:

- Runtime code: Java 5
- Orbit 4.29.0 (target platform): Java 8

Then there are some inconsistencies:

In ZipEditor MANIFEST.MF wie have JavaSE-11
and in target-platform.target 2209 (Java-17) 

I chose JavaSE-11 when I realised that there was no MacOS aarch64 implementation with Java 1.8 support. The minimal Eclipse release which supports this is 2021-06. Therefore, the target platform should contain 2021-06 at the moment.

uvoigt commented:
@mehmet-karaman 
I will commit a change in the configuration to master which addresses the inconsistency issue:
- it modifies the target platform to depend on 2021-06 as basis
- it adds dependencies to `JUnit` and the `SWT-Bot ` to the `pom.xml` of `ZipEditor-test` 

That way the tests can depend on newer release versions of these features and the target platform can nevertheless be an older one
